README
Parachute mod for Minecraft 1.7.x (Forge version)
Authored by crackedEgg
The parachute has been redesigned for 1.7.x. There are now two types of parachutes and they have their own textures. You can add
more textures to the jar file following the naming convention of the original textures. You can also,
of course, replace the textures. If you add more just set the texture index in the config file to the
index of your texture.


ChangeLog
commit ad68de8fc4a16b38d666e66f74aab6ce5942544e
Author: m1k3s <pbrane0@gmail.com>
Date:   Thu Jun 12 10:34:17 2014 -0400

    changed packet handling to SimpleNetworkWrapper. Removed old packet
    handling files and added the new ones.

commit b2798c52df4bffe46b57dc17a2c8da82a53ff70c
Merge: d16b123 5b9151a
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Sun Jun 1 14:12:20 2014 -0400

    Merge pull request #1 from Adaptivity/patch-1

    Create ru_RU.lang

commit 5b9151a18017aff63b2d49d23dd6b7847c9746c9
Author: Anton <verybigbro@gmail.com>
Date:   Sun Jun 1 14:10:06 2014 +0000

    Create ru_RU.lang

commit d16b12334db31c9db58a18a689bc0b973675e0e4
Author: m1k3s <pbrane0@gmail.com>
Date:   Tue May 27 12:07:21 2014 -0400

    I know I must have changed something....

commit 2a1f065d05a9306d205b9c3bd42e275b813a6d92
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon May 5 09:35:02 2014 -0400

    Changed dismount function from mountEntity() to
    EntityParachute.dropParachute(). shamelessly copied mountEntity code
    and modified to suit.

commit 51f0b8e1b660624bf9c9512108860edcc3025a62
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon May 5 08:14:21 2014 -0400

    Stripped out more boat shit. Oh yeah, I used the boat as a starting
    point for the parachute. I'm finding more and more code that isn't
    used in the parachute.

commit 81d1311ed89aadedce58b4fe6d8d4619c2bf0682
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat May 3 12:59:47 2014 -0400

    adjust parachute deployment 'Y' coord.

commit 97a82102a156726f940382f475b93cf4c74673e0
Author: m1k3s <pbrane0@gmail.com>
Date:   Fri May 2 08:03:19 2014 -0400

    Using Forge LivingFallEvent means I can remove some fall checking
    code in EntityParachute.

commit 478699818b9ea6c75dea5cb6eaeba2658f9e5696
Author: m1k3s <pbrane0@gmail.com>
Date:   Thu May 1 16:14:34 2014 -0400

    More tweaking on the fall event code. Eventually I may get it working
    like I think it should.

commit 84e11c4d8aa5d542c784ce80b8cf9ff7d31418da
Author: m1k3s <pbrane0@gmail.com>
Date:   Thu May 1 10:55:46 2014 -0400

    Guess I have to add the PlayerFallEvent file to git.

commit 31f6f14dff855160ffb2eae146cb426770a3c62b
Author: m1k3s <pbrane0@gmail.com>
Date:   Wed Apr 30 20:18:34 2014 -0400

    Negating fall damage by subscribing to LivingFallDamage and canceling
    if player is on a parachute.

commit 385af7bd4ff5cdb27223aa09421e839d886103d2
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon Apr 28 21:42:21 2014 -0400

    initialized ascendMode to false. Fixed bug that caused parachute to
    rise when deployed until the space bar is pressed.

commit b7db7abe2045e458d193e42ca0231c214c5de398
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat Apr 19 08:43:55 2014 -0400

    fix possible rendering bug. Wrapped all GL call blocks with
    push/popmatrix.

commit b9cbae2990543973c3b78228e1cfd94c72784a32
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon Apr 14 07:58:05 2014 -0400

    Code cleanup, initial commit to new github repo.

commit 43b4afbe8847a3c3d978f971c1c4a841d09f97cb
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat Mar 29 17:03:13 2014 -0400

    further dismount (drop the chute) tweaks. This should do it.

commit 15e5804fc9d5fdd6b549d723404e5e10525c44e3
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat Mar 29 14:26:20 2014 -0400

    changed the parachute drop handling. Using the getMountedYOffset
    function to determine height above ground.

commit 273463c5d78eda63466c84cdda10d7d1be5c0809
Author: m1k3s <pbrane0@gmail.com>
Date:   Fri Mar 21 21:51:06 2014 -0400

    fix parachute color bug, magenta was misspelled. Dammit!

commit 751491a37fe14b50c3ddc74ea4be1da2e3ec88c4
Author: m1k3s <pbrane0@gmail.com>
Date:   Sun Mar 16 15:11:05 2014 -0400

    parachute code clean up.

commit 65de101c8b99abd453170e2e84fa6432e517ee78
Author: m1k3s <pbrane0@gmail.com>
Date:   Sun Mar 16 09:48:18 2014 -0400

    added a check for the space key; only send packet if the space key
    has been pressed or released.

commit 2685d37c8ce54f9ecd41e4d83c053118bc5389c2
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat Mar 15 20:12:33 2014 -0400

    removed items drops code from EntityParachute. Parachute has not
    dropped items for several versions.

commit 7888f0cb4157811df5bb3dd7bf2caa2bbd3af99d
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat Mar 15 19:48:53 2014 -0400

    Added entity id to packet handler to test for who sent packet. Removed
    player management files.

commit 5c267516bb8c484b6476fe5f5727323b3b81695d
Author: m1k3s <pbrane0@gmail.com>
Date:   Wed Mar 12 21:20:38 2014 -0400

    Fixed single player. geez!

commit 9ed977740fab5780a51a54ef62bded37e6a33460
Author: m1k3s <pbrane0@gmail.com>
Date:   Wed Mar 12 20:14:11 2014 -0400

    More server compatibility coding. Changed PlayerInfo int mode to
    boolean keypressed.

commit 8805a64c846868d247b7da987c7ff7a931d7bda2
Author: m1k3s <pbrane0@gmail.com>
Date:   Tue Mar 11 12:54:41 2014 -0400

    Fixing major issues with server compatibility. 90% done. This works
    so far. Have not tested with other players on server.

commit 1c20737ea6ed6d3c5eea6575a90277bd80b356bc
Author: m1k3s <pbrane0@gmail.com>
Date:   Wed Mar 5 16:50:33 2014 -0500

    added:
		FMLClientHandler.instance().setDefaultMissingAction(FMLMissingMappingsEvent.Action.IGNORE);
		to FMLInitializationEvent handler. This allows the mod
		to load when there are missing item|block mappings.

commit 128645709bf7f8c5b1cae4dc3d0606949c8e6d1e
Author: m1k3s <pbrane0@gmail.com>
Date:	Mon Mar 3 17:38:57 2014 -0500

    removed assets/parachutemod/textures/blocks/*

commit c6f87e15917a748f40fc3c4a7db91cb4586e8eaf
Author: m1k3s <pbrane0@gmail.com>
Date:	Mon Mar 3 16:42:51 2014 -0500

    Changed parachute texture handling. Now the parachute uses textures
    from the current resource pack. You can set the texture color in
    the config file and as well as use the 'random' keyword.

commit b696580a0e749bb3f12c3cca2ea50ed9c1302a98
Author: m1k3s <pbrane0@gmail.com>

parachute/common/Parachute.java
//
//  =====GPL=============================================================
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; version 2 dated June, 1991.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program;  if not, write to the Free Software
//  Foundation, Inc., 675 Mass Ave., Cambridge, MA 02139, USA.
//  =====================================================================
//
//
// Copyright 2011-2014 Michael Sheppard (crackedEgg)
//
package com.parachute.common;

import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.common.MinecraftForge;
import net.minecraft.item.Item;
import net.minecraft.item.Item.ToolMaterial;
import net.minecraft.item.ItemArmor.ArmorMaterial;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.registry.*;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraftforge.common.config.Configuration;
import net.minecraftforge.common.util.EnumHelper;


@Mod(
		modid = Parachute.modid,
		name = Parachute.name,
		version = Parachute.mcversion
)

public class Parachute {
	
	public static final String modid = "parachutemod";
	public static final String mcversion = "1.8.0";
//	public static final String channel = modid;
	public static final String name = "Parachute Mod";
	public static final String parachuteName = "parachute";
	public static final String ripcordName = "ripcord";
	public static final String aadName = "auto_activation_device";
	public static final String hopnpopName = "hop_and_pop";

	static ArmorMaterial NYLON = EnumHelper.addArmorMaterial("nylon", "", 15, new int[] {2, 5, 4, 1}, 12); // same as CHAIN
	static ToolMaterial RIPSTOP = EnumHelper.addToolMaterial("ripstop", 0, 59, 2.0F, 0, 15); // same as WOOD

	private String type = parachuteName; // defaults to the normal parachute
	private boolean singleUse = false; // applies to the hop and pop chute only
	private int heightLimit = 256;
	private String chuteColor = "random";
	private boolean thermals = true;
	private double AADAltitude = 15.0;
	private boolean smallCanopy = true;
	private boolean autoDismount = true;
	private double fallThreshold = 5.0;
	private boolean weatherAffectsDrift;
	private boolean lavaThermals;
	private double minLavaDistance;
	private double maxLavaDistance;
	private boolean allowTurbulence;
	private boolean showContrails;
	private final int entityID = EntityRegistry.findGlobalUniqueEntityId();
	private static final int armorType = 1; // armor type: 0 = helmet, 1 = chestplate, 2 = legs,       3 = boots
	public static final int armorSlot = 2;  // armor slot: 0 = ??,     1 = ??,         2 = chestplate, 3 = ??

	@SidedProxy(
			clientSide = "com.parachute.client.ClientProxyParachute",
			serverSide = "com.parachute.common.CommonProxyParachute"
	)
	public static CommonProxyParachute proxy;

	public static ItemParachute parachuteItem;
	public static ItemHopAndPop hopnpopItem;
	public static ItemRipCord ripcordItem;
	public static ItemAutoActivateDevice aadItem;

	@Mod.Instance(modid)
	public static Parachute instance;

	@EventHandler
	public void preInit(FMLPreInitializationEvent event)
	{
		String generalComments = Parachute.name + " Config\nMichael Sheppard (crackedEgg)"
				+ " For Minecraft Version " + Parachute.mcversion + "\n";
		String usageComment = "set to true for hop-n-pop single use (false)";
		String heightComment = "0 (zero) disables altitude limiting (256)";
		String thermalComment = "true|false enable/disable thermals (true)";
		String lavaThermalComment = "use lava heat to get thermals to rise up, disables space bar thermals (false)";
		String minLavaDistanceComment = "minimum distance from lava to grab thermals, if you\n"
				+ "go less than 3.0 you will most likely dismount in the lava! (3.0)";
		String maxLavaDistanceComment = "maximum distance to rise from lava thermals (48)";
		String aadAltitudeComment = "altitude (in meters) at which auto deploy occurs (10)";
		String fallThresholdComment = "player must have fallen this far to activate AAD (5.0)";
		String typeComment = "set to true to use the smaller 3 panel canopy, false for the\nlarger 4 panel canopy (true)";
		String autoComment = "If true the parachute will dismount the player automatically,\n"
				+ "if false the player has to use LSHIFT to dismount the parachute. (true)";
		String weatherComment = "set to false if you don't want the drift rate to be affected by bad weather (true)";
		String turbulenceComment = "set to true to feel the turbulent world of Minecraft (false)";
		String trailsComment = "set to true to show contrails from parachute (false)";
		String colorComment = "Parachute Colors Allowed:\n"
				+ "black\nblue\n"
				+ "brown\ncyan\n"
				+ "gray\ngreen\n"
				+ "light_blue\nlime\n"
				+ "magenta\norange\n"
				+ "pink\npurple\n"
				+ "red\nsilver\n"
				+ "white\nyellow\n"
				+ "random - allows randomly chosen color each time chute is opened\n"
				+ "custom[0-9] - allows use of a custom texture called 'custom' with a single number appended";

		Configuration config = new Configuration(event.getSuggestedConfigurationFile());
		config.load();

		singleUse = config.get(Configuration.CATEGORY_GENERAL, "singleUse", false, usageComment).getBoolean(false);
		heightLimit = config.get(Configuration.CATEGORY_GENERAL, "heightLimit", 256, heightComment).getInt();
		thermals = config.get(Configuration.CATEGORY_GENERAL, "allowThermals", true, thermalComment).getBoolean(true);
		lavaThermals = config.get(Configuration.CATEGORY_GENERAL, "lavaThermals", false, lavaThermalComment).getBoolean(false);
		minLavaDistance = config.get(Configuration.CATEGORY_GENERAL, "minLavaDistance", 3.0, minLavaDistanceComment).getDouble(3.0);
		maxLavaDistance = config.get(Configuration.CATEGORY_GENERAL, "maxLavaDistance", 48.0, maxLavaDistanceComment).getDouble(48.0);
		fallThreshold = config.get(Configuration.CATEGORY_GENERAL, "fallThreshold", 5.0, fallThresholdComment).getDouble(5.0);
		AADAltitude = config.get(Configuration.CATEGORY_GENERAL, "AADAltitude", 15.0, aadAltitudeComment).getDouble(15.0);
		smallCanopy = config.get(Configuration.CATEGORY_GENERAL, "smallCanopy", true, typeComment).getBoolean(true);
		autoDismount = config.get(Configuration.CATEGORY_GENERAL, "autoDismount", true, autoComment).getBoolean(true);
		chuteColor = config.get(Configuration.CATEGORY_GENERAL, "chuteColor", "random", colorComment).getString();
		weatherAffectsDrift = config.get(Configuration.CATEGORY_GENERAL, "weatherAffectsDrift", true, weatherComment).getBoolean(true);
		allowTurbulence = config.get(Configuration.CATEGORY_GENERAL, "allowTurbulence", false, turbulenceComment).getBoolean(false);
		showContrails = config.get(Configuration.CATEGORY_GENERAL, "showContrails", false, trailsComment).getBoolean(false);
		
		// if using lava thermals disable space bar thermals, clamp the minimum lava distance.
		if (lavaThermals) {
			thermals = false;
			minLavaDistance = minLavaDistance < 2.0 ? 2.0 : minLavaDistance;
		}

		config.addCustomCategoryComment(Configuration.CATEGORY_GENERAL, generalComments);

		config.save();

		// clamp the fallThreshold to a minimum of 2
		fallThreshold = fallThreshold < 2.0 ? 2.0 : fallThreshold;

		int chuteID = proxy.addArmor(parachuteName.toLowerCase());
		EntityRegistry.registerModEntity(EntityParachute.class, parachuteName, entityID, this, 64, 10, true);
		
		// create new items, set unlocalized names and register
		parachuteItem = (ItemParachute) (new ItemParachute(NYLON, chuteID, armorType)).setUnlocalizedName(parachuteName);
		GameRegistry.registerItem(parachuteItem, parachuteName);

		ripcordItem = (ItemRipCord) (new ItemRipCord()).setUnlocalizedName(ripcordName);
		GameRegistry.registerItem(ripcordItem, ripcordName);

		aadItem = (ItemAutoActivateDevice) (new ItemAutoActivateDevice()).setUnlocalizedName(aadName);
		GameRegistry.registerItem(aadItem, aadName);

		hopnpopItem = (ItemHopAndPop) (new ItemHopAndPop(RIPSTOP)).setUnlocalizedName(hopnpopName);
		GameRegistry.registerItem(hopnpopItem, hopnpopName);

		PacketHandler.init();
	}

	@EventHandler
	public void Init(FMLInitializationEvent event)
	{
		// recipes to craft the parachutes, ripcord and AAD
		GameRegistry.addRecipe(new ItemStack(parachuteItem, 1), new Object[] {
			"###", "X X", " L ", '#', Blocks.wool, 'X', Items.string, 'L', Items.leather
		});

		GameRegistry.addRecipe(new ItemStack(hopnpopItem, 1), new Object[] {
			"###", "X X", " X ", '#', Blocks.wool, 'X', Items.string
		});

		GameRegistry.addRecipe(new ItemStack(ripcordItem, 1), new Object[] {
			"#  ", " # ", "  *", '#', Items.string, '*', Items.iron_ingot
		});

		GameRegistry.addRecipe(new ItemStack(aadItem, 1), new Object[] {
			" * ", " % ", " # ", '*', Items.comparator, '%', Items.redstone, '#', ripcordItem});

		proxy.registerRenderer();

		proxy.registerHandlers();

		FMLCommonHandler.instance().bus().register(new AADTick());
		MinecraftForge.EVENT_BUS.register(new PlayerFallEvent());
		
		proxy.registerResources();

		instance = this;
	}

	@EventHandler
	public void postInit(FMLPostInitializationEvent event)
	{
		proxy.info("ParachuteMod initialization is complete.");
	}

	public String getVersion()
	{
		return Parachute.mcversion;
	}

	public double getMaxAltitude()
	{
		return heightLimit;
	}

	public boolean getAllowThermals()
	{
		return thermals;
	}

	public String getChuteColor()
	{
		return chuteColor;
	}

	public double getAADAltitude()
	{
		return AADAltitude;
	}

	public double getFallThreshold()
	{
		return fallThreshold;
	}
	
	public boolean getAllowLavaThermals()
	{
		return lavaThermals;
	}
	
	public boolean getWeatherAffectsDrift()
	{
		return weatherAffectsDrift;
	}
	
	public double getMinLavaDistance()
	{
		return minLavaDistance;
	}
	
	public double getMaxLavaDistance()
	{
		return maxLavaDistance;
	}
	
	public boolean getAllowturbulence()
	{
		return allowTurbulence;
	}
	
	public boolean getShowContrails()
	{
		return showContrails;
	}

	public boolean isSmallCanopy()
	{
		return smallCanopy;
	}
	
	public boolean isAutoDismount()
	{
		return autoDismount;
	}

	public void setType(String type)
	{
		this.type = type;
		// force a small canopy for the hop-n-pop chute
		if (this.type.equals(hopnpopName)) {
			smallCanopy = true;
		}
	}

	public String getType()
	{
		return this.type;
	}

	public int getHopAndPopDamageAmount()
	{
		if (singleUse) {
			return hopnpopItem.getMaxDamage() + 1;
		}
		return 1;
	}

	public static boolean playerIsWearingParachute(EntityPlayer player)
	{
		ItemStack stack = player == null ? null : player.getCurrentArmor(armorSlot);
		if (stack != null) {
			Item item = stack.getItem();
			if (item != null && item instanceof ItemParachute) {
				return true;
			}
		}
		return false;
	}

	public static boolean isFalling(EntityPlayer entity)
	{
		return (entity.fallDistance > 0.0F && !entity.onGround && !entity.isOnLadder());
	}
	
}
